compiler.go

```diff
@@ -385,12 +403,30 @@ func (t *task) asFile(ctx context.Context, name string, r SearchResult) (linker.
 	}
 
 	var deps []linker.File
-	if len(parseRes.FileDescriptorProto().Dependency) > 0 {
-		t.r.setBlockedOn(parseRes.FileDescriptorProto().Dependency)
+	fileDescriptorProto := parseRes.FileDescriptorProto()
+	var wantsDescriptorProto bool
+	imports := fileDescriptorProto.Dependency
+
+	if t.e.hasOverrideDescriptorProto() {
+		// we only consider implicitly including descriptor.proto if it's overidden
+		includesDescriptorProto := name == descriptorProtoPath
+		for _, dep := range fileDescriptorProto.Dependency {
+			if !includesDescriptorProto && dep == descriptorProtoPath {
+				includesDescriptorProto = true
+			}
+		}
```

Comment:
Small optimization here, but can't you do this to stop searching once you've found an import?

```go
includesDescriptorProto := false
if name == descriptorProtoPath {
  includesDescriptorProto = true
} else {
  for _, dep := range fileDescriptorProto.Dependency {
     if dep == descriptorProtoPath {
       includesDescriptorProto = true
       break
     }
  }
}
```

Reply:
Indeed. I originally had this logic inside the loop below, which is why it couldn't break. But then I factored out, and I didn't catch this opportunity. Good catch!

Reply:
done